### ISSUE TITLE
exceeded maximum number of zone rulesets for phase 'http_ratelimit'

### ISSUE THREAD

**Issue by fivesheep** [cloudflare/terraform-provider-cloudflare#2634]
### Confirmation

- [X] My issue isn't already found on the issue tracker.
- [X] I have replicated my issue using the latest version of the provider and it is still present.

### Terraform and Cloudflare provider version

terravorm version 1.4.6
cloudflare provider: 4.10.0

### Affected resource(s)

http_ratelimit

### Terraform configuration files

```hcl
locals {
  rate_limits = {
    get = {

    }
    post = {

    }
  }
}

resource "cloudflare_ruleset" "this" {
  for_each = locals.rate_limits

  zone_id     = "<MY-ZONE-ID>"
  name        = each.key
  description = "desc for ${each.key}"
  kind        = "zone"
  phase       = "http_ratelimit"

  rules {
    action = "log"

    ratelimit {
      characteristics = [
        "cf.colo.id",
        "ip.src"
      ]
      period              = 60
      requests_per_period = 10
      mitigation_timeout  = 60
      requests_to_origin  = true
    }

    expression  = "<some exporession>"
    description = "rate limit for ${each.key}"
    enabled     = true
  }
}
```

### Link to debug output

N/A

### Panic output

```
╷
│ Error: error creating ruleset api_write
│ 
│   with module.rate_limit_rules["api_write"].cloudflare_ruleset.this,
│   on modules/cloudflare/rate-limit/main.tf line 2, in resource "cloudflare_ruleset" "this":
│    2: resource "cloudflare_ruleset" "this" {
│ 
│ exceeded maximum number of zone rulesets for phase 'http_ratelimit'
╵
```

### Expected output

create two ratelimit rules

### Actual output

```
│ exceeded maximum number of zone rulesets for phase 'http_ratelimit'
```

### Steps to reproduce

1. clear ratelimit rules from security->waf ui, confirm with api no http_ratelimit rule defined, and ui showing 0 out of 100 rule was used
2. plan and apply terraform
3. one of the rule got created, the other failed with the error
4. manual create the other rule from ui worked
5. check api, no other rule for this was creeated

### References

https://github.com/cloudflare/terraform-provider-cloudflare/issues/2004  
there's a simliar issue on a different rule was reported, but with the provide approach, I could confirm we didn't have anything defined from api response

**Comment by jacobbednarz** [comment]
you cannot have duplicate phases - see https://developers.cloudflare.com/ruleset-engine/about/. you have multiple rules within a single phase.

**Comment by fivesheep** [comment]
@jacobbednarz if I understand this correctly, for each zone we can only have one ruleset per phase, and for all the rate limit rules, we can only define with one single `cloudflare_ruleset` resource?  that would make terraform very hard to manage, we are aiming to create reusable modules and hand over to others to define their rules from their applicable modules. why did we design the resource this way, is there a way we can offload the rule creation?

**Comment by jacobbednarz** [comment]
yes that is correct. the terraform resource matches the service and it was designed that way as you cannot have multiple execution points for rate limiting -- there is only the single phase of it.

**Comment by carlpett** [comment]
@jacobbednarz Looking at the API docs for the ruleset, it seems the API can add/update/remove a single rule in a ruleset, rather than always working with the full set of rules. Would it be feasible to have a `cloudflare_ruleset_rule` resource which manages individual rules?

**Comment by carlpett** [comment]
Apologies if it came across as "can't you just", I did not mean to intend it would be easy, but whether it was possible. And apparently I searched poorly since I didn't see that this had been suggested many times before...

As mentioned elsewhere, this does make it quite hard for large-scale customers to work with this, so we'd welcome investments here. But given your replies in the other threads on the topic, I do see the challenges.
